sentry-core/src/main/java/io/sentry/core/EnvelopeSender.java

```diff
@@ -0,0 +1,107 @@
+package io.sentry.core;
+
+import io.sentry.core.util.Objects;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.Charset;
+import org.jetbrains.annotations.NotNull;
+
+public final class EnvelopeSender implements IEnvelopeSender {
+  private static final Charset UTF_8 = Charset.forName("UTF-8");
+
+  private final IHub hub;
+  private final IEnvelopeReader envelopeReader;
+  private final ISerializer serializer;
+  private final ILogger logger;
+
+  public EnvelopeSender(
+      IHub hub, IEnvelopeReader envelopeReader, ISerializer serializer, ILogger logger) {
+    this.hub = Objects.requireNonNull(hub, "Hub is required.");
+    this.envelopeReader = Objects.requireNonNull(envelopeReader, "Envelope reader is required.");
+    this.serializer = Objects.requireNonNull(serializer, "Serializer is required.");
+    this.logger = Objects.requireNonNull(logger, "Logger is required.");
+  }
+
+  @Override
+  public void processEnvelopeFile(@NotNull String path) {
+    InputStream stream = null;
+    File file = null;
+    try {
+      file = new File(path);
+      stream = new FileInputStream(file);
+      SentryEnvelope envelope = envelopeReader.read(stream);
+      if (envelope == null) {
+        logger.log(SentryLevel.ERROR, "Stream from path %s resulted in a null envelope.", path);
+      } else {
+        processEnvelope(envelope);
+      }
+    } catch (Exception e) {
+      logger.log(SentryLevel.ERROR, "Error processing envelope.", e);
+    } finally {
+      try {
+        if (stream != null) stream.close();
```

Comment:
you may use `try-with-resources`

Reply:
Don't think we can use `try-with-resource` is API 19 or higher

Reply:
`In addition to the Java 8 language features and APIs above, Android Studio 3.0 and later extends support for try-with-resources to all Android API levels.`
ref https://developer.android.com/studio/write/java8-support

`AutoCloseable` is min API 19, try-with-resources is fine afaik

Reply:
OK, @metlos had to get rid of some try-with-resource on his HTTP PR, but I guess at the time we didn't know about this desugaring (I still didn't know).